### PR TITLE
Move dmd.mtype.TypeFunction.parameterStorageClass out of dmd/mtype.d

### PR DESCRIPTION
**Pull request by RazvanN7**
Because it calls determineFields which requires semantic analysis.

@ibuclaw I have erased this function from the C++ headers as it appears to be unused on the GDC side (it is used by the escape analysis module which does not have any `extern(C++)` entrypoints). However, if GDC or LDC (cc @kinke ) use it, I can make it `extern(C++)` again.

### CODE REVIEW

File: compiler/src/dmd/mtype.d [hunk 1]
```diff
@@ -4115,6 +4115,7 @@ extern (C++) final class TypeFunction : TypeNext
         return linkage == LINK.d && parameterList.varargs == VarArg.variadic;
     }
 
+<<<<<<< HEAD
```
Comment (ibuclaw) [inline compiler/src/dmd/mtype.d:4118]: :-)

Comment (RazvanN7) [reply compiler/src/dmd/mtype.d]: yep, rebase screw up. All good now.

File: compiler/src/dmd/typesem.d [hunk 2]
```diff
@@ -5541,6 +5541,122 @@ Lnotcovariant:
...
+ * (This will not affect the name mangling.)
+ * Params:
+ *  tthis = type of `this` parameter, null if none
+ *  p = parameter to this function
+ *  outerVars = context variables p could escape into, if any
+ *  indirect = is this for an indirect or virtual function call?
+ * Returns:
+ *  storage class with STC.scope_ or STC.return_ OR'd in
+ */
+StorageClass parameterStorageClass(TypeFunction tf, Type tthis, Parameter p, VarDeclarations* outerVars = null,
```
Comment (thewilsonator) [inline compiler/src/dmd/typesem.d:5557]: `src/dmd/typesem.d(5557): Warning: Ddoc arameter count mismatch, expected 5, got 4` warnings are errors

Comment (RazvanN7) [reply compiler/src/dmd/typesem.d:5558]: Solved. Thanks!

**Review by thewilsonator — APPROVED** [review]

**Pull request merged**